### PR TITLE
ccloud: force promote replica

### PR DESCRIPTION
**Pull request by chuan137**
This pull request introduces a force option to the share-replica-promote API. Normally, Manila drivers should verify that data replication is in sync with the source before promoting a replica to prevent data loss. Nevertheless, there are situations where promotion may proceed despite the status of the replication. The inclusion of this option in the API enables users to control the behavior of the driver. Additionally, it enhances verification in the NetApp driver by inspecting the health of SnapMirror before breaking it.

Change-Id: Ia08caf6d3f5864a112ce402567eddbd0d0aa1819

### CODE REVIEW

File: manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py [hunk 1]
```diff
@@ -3124,6 +3124,8 @@ def promote_replica(self, context, replica_list, replica, access_rules,
                 constants.STATUS_ERROR)
             new_active_replica['status'] = constants.STATUS_ERROR
             return [new_active_replica]
+        except exception.ReplicationException as e:
+            LOG.exception("Could not promote replica %s. %s", replica['id'], e)
```
Comment (Carthaca) [inline manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py:3128]: I think this should get the same treatment like the `exception.StorageCommunicationException` which is caught directly above, i.e. setting the replica to error and return

File: manila/share/drivers/netapp/dataontap/cluster_mode/data_motion.py [hunk 2]
```diff
@@ -369,7 +372,37 @@ def wait_for_quiesced():
                                              dest_volume,
                                              clear_checkpoint=False)
 
-    def break_snapmirror(self, source_share_obj, dest_share_obj, mount=True):
+    def check_snapmirror_health(self, source_share_obj, dest_share_obj):
+        """Check if SnapMirror relationship is in sync."""
+        snapmirror = self.get_snapmirrors(source_share_obj, dest_share_obj)[0]
+        if snapmirror.get('relationship-status') != 'idle':
```
Comment (Carthaca) [inline manila/share/drivers/netapp/dataontap/cluster_mode/data_motion.py:378]: I think the expected state in happy path would be `quiesced`.
If quiesce did not finish in time and it was aborted, it goes into `idle`, I believe.
So we should accept both values for the relationship status.

**Review by Carthaca — CHANGES_REQUESTED** [review]
With manila upgrade we will also get https://github.com/openstack/manila/commit/5f0f14a7f8b99ac008be4604bcaeec2dbc149d06, but this is useful on top of that 👍 

See my inline comments for requested changes, please

File: manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py (hunk 1 above)
Comment (chuan137) [reply manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py:3128]: any idea how to populate the error to user in driver, we have to use the message api?

File: manila/share/drivers/netapp/dataontap/cluster_mode/data_motion.py (hunk 2 above)
Comment (chuan137) [reply manila/share/drivers/netapp/dataontap/cluster_mode/data_motion.py]: corrected

**Comment by chuan137** [comment]
thanks for the review, I agree that the exception handling needs improvement. It could work to return the replica with error states as you suggested. I think the states are then updated here, https://github.com/sapcc/manila/blob/4daad526b93b90f99b046c02a7ce24ee8bb02095/manila/share/manager.py#L2794-L2816
 It's a bit clumsy to read, and in the end it always log "promoted to active state successfully", which is not always true, I think. So I try to handle the unhealthy error explicitly https://github.com/sapcc/manila/blob/4daad526b93b90f99b046c02a7ce24ee8bb02095/manila/share/manager.py#L2699C1-L2714C33
 the to-be-promoted replica will be in error states, (other replicas are not touched), and a message is sent so the reason is more transparent to the user.

**Review by Carthaca — DISMISSED** [review]
LGTM

**Review by Carthaca — APPROVED** [review]

File: manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py [hunk 3]
```diff
@@ -3324,8 +3324,12 @@ def _convert_destination_replica_to_independent(
             LOG.warning(msg)
             pass
         # 2. Break SnapMirror
-        dm_session.break_snapmirror(orig_active_replica, replica,
-                                    check_health=True)
+        if force:
```
Comment (kpawar-sap) [inline manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py]: dm_session.break_snapmirror(orig_active_replica, replica,
                                    check_health=False if force else True)

**Review by kpawar-sap — APPROVED** [review]

**Comment by Carthaca** [comment]
on newer releases we can use the upstream `quiesce_wait_time` parameter and have no need for a custom client patch

**Pull request merged**